### PR TITLE
Adding derive() macro support to SharedStructs

### PR DESCRIPTION
**Pull request by rkreutz**
This PR adds support for `#[derive(...)]` macros to `SharedStruct`s.

## Example

```rust
#[swift_bridge::bridge]
mod ffi {
    #[swift_bridge(swift_repr = "struct")]
    #[derive(Copy, Clone)]
    struct Foo {
        field: u8,
    }
}
```

This will successfully add a derived `Copy` and `Clone` trait to the struct declared.

### CODE REVIEW

File: crates/swift-bridge-ir/src/codegen/generate_rust_tokens/shared_struct.rs
Comment (chinedufn) [inline crates/swift-bridge-ir/src/codegen/generate_rust_tokens/shared_struct.rs:93]: We're moving away from automatic derives, so `derives` is fine.

File: crates/swift-bridge-ir/src/bridged_type/shared_struct.rs
Comment (chinedufn) [inline crates/swift-bridge-ir/src/bridged_type/shared_struct.rs:22]: Instead of `Option<Vec<TokenStream>>` let's use something like:

```rust
pub(crate) struct StructDerives {
    pub copy: bool,
    pub clone: bool,
}
```

File: crates/swift-bridge-ir/src/parse/parse_struct.rs
Comment (chinedufn) [inline crates/swift-bridge-ir/src/parse/parse_struct.rs:349]: Can we remove this derive?

File: crates/swift-bridge-ir/src/parse/parse_struct.rs
Comment (chinedufn) [inline crates/swift-bridge-ir/src/parse/parse_struct.rs:329]: Nice

**Review by chinedufn — COMMENTED** [review]
Nice looks great. Left a couple minor comments then we can land.

**Comment by chinedufn** [comment]
Just missing a codegen test and an integration test (don't need the Swift side .. just a Rust integration test is fine for now and in the future we can emit a clone function on the Swift side...)

See the following files in https://github.com/chinedufn/swift-bridge/pull/194/files for an example that you can copy/paste/modify

- `crates/swift-bridge-ir/src/codegen/codegen_tests/derive_attribute_codegen_tests.rs`
- `crates/swift-integration-tests/src/enum_attributes/derive.rs`

File: crates/swift-bridge-ir/src/bridged_type/shared_struct.rs
Comment (rkreutz) [reply crates/swift-bridge-ir/src/bridged_type/shared_struct.rs:22]: This would kind of limit us to only these kind of derives, if we want to add new ones (like `Debug`) we would need to open a PR and add them. It also has the potential of growing quite a lot. I don't see why it would be an issue to have the more generic `Vec<TokenStream>`?

File: crates/swift-bridge-ir/src/parse/parse_struct.rs
Comment (rkreutz) [reply crates/swift-bridge-ir/src/parse/parse_struct.rs:349]: I'm actually testing here whether `#[swift_bridge]` and `#[derive]` work together here. I think it's a good test to have, so I'll just leave this test as is and duplicate it to check if they work ok together

File: crates/swift-bridge-ir/src/bridged_type/shared_struct.rs
Comment (chinedufn) [reply crates/swift-bridge-ir/src/bridged_type/shared_struct.rs:22]: Thanks for explaining your thinking. Makes sense.

---

The issue is that we only want to support derives that we know are well behaved. 
Supporting arbitrary derives might be possible, but requires more research.

As it stands now, supporting arbitrary derives can lead to undefined behavior. 
It's incredibly unlikely, but theoretically possible. 

Before we support arbitrary derives we want to make UB *completely* impossible.

---

Essentially, `#[derive(SomeCustomDerive)]` could emit something like `type u32 = some_crate::NotAU32` and cause `swift-bridge` to think that it sees a `u32` not realizing that a type alias got generated to make `u32` a completely different type.

It should be possible to guard against this, but until we guard against it (and any other potential derive related UB scenarios) we can't allow arbitrary derives.

This discussion has more detail https://github.com/chinedufn/swift-bridge/issues/190#issuecomment-1462315921

**Comment by rkreutz** [comment]
> See the following files in https://github.com/chinedufn/swift-bridge/pull/194/files for an example that you can copy/paste/modify

> crates/swift-bridge-ir/src/codegen/codegen_tests/derive_attribute_codegen_tests.rs

@chinedufn should I wait for that PR to be merged to add the tests on this file you mentioned, or should I create a separate file like `derive_struct_attribute_codegen_tests.rs`?

File: crates/swift-bridge-ir/src/bridged_type/shared_struct.rs
Comment (rkreutz) [reply crates/swift-bridge-ir/src/bridged_type/shared_struct.rs:22]: makes sense, will go with your suggestion then

File: crates/swift-bridge-ir/src/parse/parse_struct.rs
Comment (rkreutz) [reply crates/swift-bridge-ir/src/parse/parse_struct.rs:392]: [ead0af9](https://github.com/chinedufn/swift-bridge/pull/198/commits/ead0af99013d5cafb5e2cb8d0da6ba7a191d8211)

File: crates/swift-bridge-ir/src/bridged_type/shared_struct.rs
Comment (rkreutz) [reply crates/swift-bridge-ir/src/bridged_type/shared_struct.rs]: [ead0af9](https://github.com/chinedufn/swift-bridge/pull/198/commits/ead0af99013d5cafb5e2cb8d0da6ba7a191d8211)

File: crates/swift-bridge-ir/src/codegen/generate_rust_tokens/shared_struct.rs
Comment (rkreutz) [reply crates/swift-bridge-ir/src/codegen/generate_rust_tokens/shared_struct.rs]: [ead0af9](https://github.com/chinedufn/swift-bridge/pull/198/commits/ead0af99013d5cafb5e2cb8d0da6ba7a191d8211)

File: crates/swift-bridge-ir/src/parse/parse_struct.rs
Comment (chinedufn) [inline crates/swift-bridge-ir/src/parse/parse_struct.rs:361]: nice

File: crates/swift-integration-tests/src/struct_attributes/derive.rs
Comment (chinedufn) [inline crates/swift-integration-tests/src/struct_attributes/derive.rs:1]: Awesome tests. Might've been nice to actually use the `Copy/Clone` impls outside of the module down below, but this is fine.

**Review by chinedufn — APPROVED** [review]
Looks great! Thanks!!

**Comment by chinedufn** [comment]
Creating a file in this PR sounds good!

**Comment by rkreutz** [comment]
> Creating a file in this PR sounds good!

just updated with the tests

**Comment by chinedufn** [comment]
Looks good. I'll merge once tests are passing (looks like you need to run `cargo fmt`)

**Comment by chinedufn** [comment]
Thank you!

**Comment by extrawurst** [comment]
This would be great to have on `enum` too

**Pull request merged**
